Scale-invariant levenberg marquardt

michael-0brien opened the pull request:
Here's my draft of the `AbstractDescent` as described in #187. I also have the inclusion of an initial step size to the `ClassicalTrustRegion` here, but we can move this to another PR if desired.

For making scale-invariance the default implementation, we will have to discuss how to strike a balance between scale-invariance and robustness to bad initialization. The jist is that if the initial hessian estimate is bad, the algorithm loses parameter sensitivity.

The scipy version involves taking the scaling matrix `D` to be the max diagonal of the hessian so far encountered. They probably have other tricks, but this alone is not robust to bad initialization. I could imagine other cases where one would not want the max, perhaps one would be interested in the average instead. Regardless, we will need some modularity for implementing the scaling operator.

The paper I shared in #187 has some advice for choices of `D`, such as adding small const diagonal offset to prevent becoming too insensitive to parameters. In my work, I'm having success normalizing the trace of `D` to be equal to the identity. That way the LM param too carries a sense of scale and I can use it to bias the algorithm to gradient-like or newton-like steps in early iterations. `D` then maintains relative scale between parameters. I'm not suggesting this should be the default, but some food for thought.

michael-0brien commented:
Two things added to public API so far:

- `ScaledDampedNewtonDescent`: The `AbstractDescent` for direct scale-invariant LM. This keeps track of the scaling operator and allows a user to pass a callable `update_scaling_fn(hessian: lx.AbstractLinearOperator, scaling_operator: lx.DiagonalLinearOperator) -> lx.DiagonalLinearOperator`.
- `max_diagonal_scaling_update`: The "scipy" scaling update of taking the maximum diagonal of the hessian so far encountered. I thought that exposing scaling update functions to public API could follow functions for computing beta in non-linear CG.

Not sure if this is fully the way to go but should get the ball rolling! I've also written up some documentation and added to tests.

johannahaffner commented on optimistix/_solver/levenberg_marquardt.py:
```diff
@@ -31,42 +37,79 @@ def __call__(self, y: PyTree[Array]):
+    hessian: lx.AbstractLinearOperator, scaling_operator: lx.DiagonalLinearOperator
+) -> lx.DiagonalLinearOperator:
+    """Update the matrix `D` that controls the relative scaling of each
+    parameter based on the procedure described by More (1977).
```
Can we make that reference more specific?

johannahaffner commented on optimistix/_solver/levenberg_marquardt.py:
```diff
@@ -196,7 +239,10 @@ def query(
         newton_norm = self.trust_region_norm(newton)
         return _IndirectDampedNewtonDescentState(
-            f_info=f_info, newton=newton, newton_norm=newton_norm, result=result
+            f_info=f_info,
```
This looks like a spurious formatting change - best to remove it, I think.

johannahaffner commented on optimistix/_solver/levenberg_marquardt.py:
```diff
@@ -31,42 +37,79 @@ def __call__(self, y: PyTree[Array]):
+    scaling_operator: lx.DiagonalLinearOperator
+
+    def __call__(self, hessian: lx.AbstractLinearOperator) -> lx.DiagonalLinearOperator:
+        return self.update_scaling_fn(hessian, self.scaling_operator)
```
I don't think this will work if `scaling_update` is `None`, which we do implicitly allow since this is one of the things we check in `damped_newton_step` below. 

I'll leave another comment at the call site.

johannahaffner commented:
> The jist is that if the initial hessian estimate is bad, the algorithm loses parameter sensitivity.

How does this play out with different flavours of Hessian approximations? A `BFGS` type Hessian approximation that starts out with an identity should basically implement gradient descent for the first few steps. What happens with a residual Jacobian far from the optimum?

> I could imagine other cases where one would not want the max, perhaps one would be interested in the average instead. 

>  In my work, I'm having success normalizing the trace of `D` to be equal to the identity. That way the LM param too carries a sense of scale and I can use it to bias the algorithm to gradient-like or newton-like steps in early iterations. 

> Regardless, we will need some modularity for implementing the scaling operator.

I think making the type of the update function more flexible should address all of the above, right? This does not answer the question of what the best default is, but we could start out with the maximum value, matching `scipy`, and make a change if we see something else that performs much better. In any case, we should write in the documentation that this can be highly problem-specific. 

We could also provide several scaling functions, similar to how we have implemented this for the `NonlinearCG` family, which may be used with different methods: https://docs.kidger.site/optimistix/api/minimise/#optimistix.NonlinearCG (that are also just restricted to implement a specific signature).

michael-0brien commented:
Thanks for the comments! I’ll work on finalizing this PR. Two questions in finalizing: 

1. What API would we like for the `LevenbergMarquardt` class? Should the scaled update become the default and if so, how? Didn’t want to touch this part since I’m not sure what is desired for backwards compatibility. Somehow it should be clear that the new approach may break for large systems since it requires materializing the hessian. 
2. Would you prefer I revert my commit adding the `initial_step_size` to the search method, or would you want to cherry pick my commit for the scaled descent? If the latter, I’ll add another commit addressing your comments.

> > The jist is that if the initial hessian estimate is bad, the algorithm loses parameter sensitivity.
> 
> How does this play out with different flavours of Hessian approximations? A `BFGS` type Hessian approximation that starts out with an identity should basically implement gradient descent for the first few steps. What happens with a residual Jacobian far from the optimum?

Good question, I need to read up on this. The work I’ve seen that studies this is always in context of the usual residual jacobian LM. I think this is still true in this case, which is why I’m curious if scipy does anything else under the hood besides what’s implemented in the 1977 reference they cite. 

I am using a BFGS hessian of course, so I just bias the algorithm in early iterations to be gradient-like. I do this by choosing the initial LM param to be large, and also by using a custom search that has the behavior of `LinearTrustRegion` before some `hessian_burn_in` number of steps, and `ClassicalTrustRegion` after that.

> > I could imagine other cases where one would not want the max, perhaps one would be interested in the average instead.
> 
> > In my work, I'm having success normalizing the trace of `D` to be equal to the identity. That way the LM param too carries a sense of scale and I can use it to bias the algorithm to gradient-like or newton-like steps in early iterations.
> 
> > Regardless, we will need some modularity for implementing the scaling operator.
> 
> I think making the type of the update function more flexible should address all of the above, right? This does not answer the question of what the best default is, but we could start out with the maximum value, matching `scipy`, and make a change if we see something else that performs much better. In any case, we should write in the documentation that this can be highly problem-specific.
> 
> We could also provide several scaling functions, similar to how we have implemented this for the `NonlinearCG` family, which may be used with different methods: https://docs.kidger.site/optimistix/api/minimise/#optimistix.NonlinearCG (that are also just restricted to implement a specific signature).

Yes, I think so. Going with the scipy version and providing flexibility is a good plan!

johannahaffner commented:
> Comments should be addressed. Added the new `ScaledLevenbergMarquardt`, `max_diagonal_scaling_update`, and `ScaledDampedNewtonDescent` to the docs.

Great, will take a look tomorrow!

johannahaffner commented on optimistix/_solver/levenberg_marquardt.py:
```diff
@@ -253,6 +280,102 @@ def accept_newton():
+def max_diagonal_scaling_update(
+    hessian: lx.AbstractLinearOperator, scaling_operator: lx.AbstractLinearOperator
+) -> lx.AbstractLinearOperator:
+    """Update the scaling matrix `D` as the maximum diagonal of the
```
Wouldn't it be desirable to regularise the Hessian less stringently as the optimisation progresses? If we keep the maximum value around forever I would expect convergence to slow down - the regulariser will be too dominant, and the system will become too gradient-like for fast convergence.

johannahaffner replied:
Ok I have more questions about this one. In the [More 1977](https://scispace.com/pdf/levenberg-marquardt-algorithm-implementation-and-theory-9p7qhokmlq.pdf) paper, equation (6.3) does seem to point to a related concept. Here we take whatever is larger between 

- the value on the diagonal of the Hessian approximation
- the norm of the respective column in the Jacobian

in an element-wise fashion, but going back just one iteration. In other words, we would not accumulate large values throughout the optimisation. What is your take on this section of the paper?

`max_diagonal_scaling_update` should also have a reference, I think, even if this ends up being a bit repetitive given that we have references elsewhere. As soon as we'll have more than one of these scaling functions in the public API the correct reference won't be immediately obvious anymore.

johannahaffner reviewed (commented):
Looks good! I have some nits, some documentation requests and a question about the implementation of the diagonal scaling update - if I read the More paper correctly then they propose something different there which does not accumulate values throughout the optimisation, but rather only between successive iterations.

michael-0brien commented on optimistix/_solver/levenberg_marquardt.py:
```diff
@@ -253,6 +280,102 @@ def accept_newton():
+
+    Taking `D = I` is the original procedure proposed by Levenberg, whereas
+    More (1977) describes taking `D` as the maximum diagonal of the hessian
+    so far encountered. This procedure is used in [`scipy.least_squares`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html)
```
My interpretation of what scipy does, and the interpretation of the More paper come from this reference: https://arxiv.org/abs/1201.5885

Perhaps should not have taken this too much at face value. I’m not quite sure what the right way to go is. But if it’s true that scipy does indeed use the max diagonal so far encountered, even if this is not what is described in More, it might be the right default to go with? Points taken though.

Based on this reference do you think we should investigate scipy further or implement what is in More? If the latter, let me know what you think the best approach would be.

michael-0brien replied:
Reading through the More paper again, I think what is described (Eq 6.3) is indeed what we’ve implemented. What makes you think the maximum is with respect to current and previous iterate?

johannahaffner replied:
In the More paper - I interpreted the index $k-1$ as the value of the diagonal element in `Jac^T Jac` from the previous iteration, but looking at it again I can also see your interpretation. Since we're defining a recursion it actually makes a lot of sense to define it the way you did. 

What I find very surprising is that it is sensible to keep the largest value _thus far encountered_ for the entirety of the optimisation procedure. That will accumulate information from the start of the optimisation to the end, and I don't find that intuitive.

I'll take another look at the Scipy code, I'm still a bit hesitant to claim equivalence.

michael-0brien replied:
Not sure if you already found this, but there is a [modernization of MINPACK](https://github.com/fortran-lang/minpack) that looks very useful for us! Here is the documentation + source code for the lmder routine, which is based off the implementation from the More paper: https://fortran-lang.github.io/minpack/proc/lmder.html

Looks like the repo is actively maintained too if we want to get in touch.

michael-0brien commented:
I’ll hold off on pushing more changes until we get a handle on the scaling update, but I’ll get these changes pushed soon after!

michael-0brien commented:
@johannahaffner @patrick-kidger where are we with this? There are some open questions, but as long as we are confident about API and we are sure it is a standard implementation in sciopt, then it seems like we can merge?

I think the only change that would be good before merge is the scaling operator is currently instantiated as `D = diag(-jnp.inf, …)`, but this disallows custom additive update rules like one with momentum, i.e. `D_{k+1} = (1 - p) D_k + (1 - p)**2 D_{k-1} + …` (see above paper; update rules like this are used for determining diagonal SGD preconditioners, which is closely related). I think it should work to instantiate it with zeros instead if the hessian diagonal is positive definite. 

No rush, just a friendly ping + wanted to pass along my updated thoughts :).
